docs/upgrade_1-2_to_1-3.md

```diff
@@ -0,0 +1,52 @@
+# Upgrading from OADP 1.2
+
+> **NOTE:** Always upgrade to next minor version, do NOT skip versions. To update to higher version, do one upgrade at a time. Example: to upgrade from 1.1 to 1.3, upgrade first to 1.2, then to 1.3.
+## Changes from OADP 1.2 to 1.3
+
+- Velero was updated from version 1.11 to 1.12 (Changes reference: https://velero.io/docs/v1.12/upgrade-to-1.12/#upgrade-from-v110-or-higher)
+
+    From this update, OADP now uses Velero Built-in DataMover instead of VSM/Volsync DataMover. This change the following:
+
+    - `spec.features.dataMover` was removed (you can delete secret for `spec.features.dataMover.credentialName` from your cluster, if you want)
+
+    - `vsm` plugin was removed
+
+    - `volsync` operator is not necessary anymore (you can uninstall it from your cluster, if you want)
+
+    - `volumesnapshotbackups.datamover.oadp.openshift.io` and `volumesnapshotrestores.datamover.oadp.openshift.io` CustomResourceDefinitions are not necessary anymore (you can delete `volumesnapshotbackups.datamover.oadp.openshift.io` and `volumesnapshotrestores.datamover.oadp.openshift.io` CRD from your cluster, if you want)
+
```

Comment:
I need to test if this is all. Do you remember anything? There is anything in especial you want to add the responsibility to OADP code to handle?

Reply:
I guess we could have OADP remove these VSB/VSR CRs. Or we could just leave this as a doc for users to clean up if they want to -- leaving them around won't break anything.

Reply:
+1 with user initiating clean up. We should automate deletion as minimally required IMO in case of downgrade.

Reply:
Also talk about restic to nodeAgent change

Reply:
tried to run an upgrade test, but it did not work :( during upgrade got this error

error validating existing CRs against new CRD's schema for "dataprotectionapplications.oadp.openshift.io": error validating custom resource against new schema for DataProtectionApplication openshift-adp/test-upgrade: [].spec.configuration.velero.defaultPlugins[3]: Unsupported value: "vsm": supported values: "aws", "gcp", "azure", "csi", "openshift", "kubevirt"

to test it locally, run `make catalog-test-upgrade` and run the `oc` command required after that

Is this a problem or we should require users to stop running instance of DPA before update? (instead of changing channel, do the steps from `docs/upgrade_1-2_to_1-3.md`)

Reply:
Relevant issue: https://issues.redhat.com/browse/OCPBUGS-15819
Once upgrade started, fixing dpa won't resolve this. You need to uninstall then reinstall.

Reply:
Alright, fixed the automatically upgrade bug in following commit

I think for leftover we have 
- volume-snapshot-mover deployment (will write code to delete it)
- Volsync operator (I would not remove it, since its cluster wide, maybe there are users that use it for other things?)
- and CRDs

Do you agree with volsync approach? How should we delete old CRDs? Can just delete, first check if tehre no running instances? if there are: tell users to delete them themselves?

Reply:
Yes please do not add logic to remove volsync operator, volsync controller, deployment and its CRDs. We do not own these, volsync operator and its resources may/can exist independently without any correlation with OADP operator.

Reply:
wrote the deletion code in `controllers/datamover.go`, it looks like this
```go
package controllers

import (
	"github.com/go-logr/logr"
	appsv1 "k8s.io/api/apps/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime/schema"
)

func (r *DPAReconciler) ReconcileDataMoverController(log logr.Logger) (bool, error) {
	// When updating from OADP version 1.2.x to 1.3.x, delete
	//   volume-snapshot-mover deployment
	//   volumesnapshotbackups.datamover.oadp.openshift.io CRD
	//   volumesnapshotrestores.datamover.oadp.openshift.io CRD
	// because OADP now uses Velero builtin DataMover instead of VSM/Volsync DataMover
	if err := r.Delete(r.Context, &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{
		Name:      "volume-snapshot-mover",
		Namespace: r.NamespacedName.Namespace,
		Labels:    map[string]string{"openshift.io/oadp-data-mover": "True"},
	}}); err != nil && !errors.IsNotFound(err) {
		return false, err
	}

	volumeSnapshotBackupsResource := &unstructured.Unstructured{}
	volumeSnapshotBackupsResource.SetGroupVersionKind(schema.GroupVersionKind{
		Group:   "apiextensions.k8s.io",
		Version: "v1",
		Kind:    "CustomResourceDefinition",
	})
	volumeSnapshotBackupsResource.SetName("volumesnapshotbackups.datamover.oadp.openshift.io")
	if err := r.Delete(r.Context, volumeSnapshotBackupsResource); err != nil && !errors.IsNotFound(err) {
		return false, err
	}

	volumeSnapshotRestoresResource := &unstructured.Unstructured{}
	volumeSnapshotRestoresResource.SetGroupVersionKind(schema.GroupVersionKind{
		Group:   "apiextensions.k8s.io",
		Version: "v1",
		Kind:    "CustomResourceDefinition",
	})
	volumeSnapshotRestoresResource.SetName("volumesnapshotrestores.datamover.oadp.openshift.io")
	if err := r.Delete(r.Context, volumeSnapshotRestoresResource); err != nil && !errors.IsNotFound(err) {
		return false, err
	}

	return true, nil
}
```

plus adding this to `config/rbac/role.yaml`
```yaml
- apiGroups:
  - apiextensions.k8s.io
  verbs:
  - delete
  resources:
  - customresourcedefinitions
  resourceNames:
  - volumesnapshotbackups.datamover.oadp.openshift.io
  - volumesnapshotrestores.datamover.oadp.openshift.io
```

Can I commit this?

Reply:
We add finalizer to VSB/VSR [here](https://github.com/search?q=repo%3Amigtools%2Fvolume-snapshot-mover%20dmFinalizer&type=code)

Reply:
We remove finalizer [here](https://github.com/search?q=repo%3Amigtools%2Fvolume-snapshot-mover%20RemoveFinalizer&type=code)

Reply:
Haven't found where we call delete for the user yet.

Reply:
As talked in scrum, we will not delete CRDs because failed backup/restores can have VSB/VSR leftovers that have finalizers, and that can be trouble

We will add a `oc` CLI commands to users that want to delete them, in another PR

resolving conversation